common/argparse_test.cpp

```diff
@@ -0,0 +1,853 @@
+                        [&](auto& arg_b) { arg_b.Append(&strings); });
+      b.Do([] {});
+    });
+  };
+
+  EXPECT_THAT(parse({}, llvm::errs()), Eq(Args::ParseResult::Success));
+  EXPECT_THAT(integers, ElementsAre());
+  EXPECT_THAT(strings, ElementsAre());
+
+  EXPECT_THAT(parse({"--int=1", "--int=2", "--int=3"}, llvm::errs()),
+              Eq(Args::ParseResult::Success));
+  EXPECT_THAT(integers, ElementsAre(Eq(1), Eq(2), Eq(3)));
+  EXPECT_THAT(strings, ElementsAre());
+
+  EXPECT_THAT(parse({"--str=a", "--str=b", "--str=c"}, llvm::errs()),
+              Eq(Args::ParseResult::Success));
+  EXPECT_THAT(integers, ElementsAre(Eq(1), Eq(2), Eq(3)));
+  EXPECT_THAT(strings, ElementsAre(StrEq("a"), StrEq("b"), StrEq("c")));
+
+  EXPECT_THAT(parse({"--str=d", "--int=4", "--str=e"}, llvm::errs()),
+              Eq(Args::ParseResult::Success));
+  EXPECT_THAT(integers, ElementsAre(Eq(1), Eq(2), Eq(3), Eq(4)));
+  EXPECT_THAT(strings, ElementsAre(StrEq("a"), StrEq("b"), StrEq("c"),
+                                   StrEq("d"), StrEq("e")));
+}
+
+TEST(ArgParserTest, PositionalAppending) {
+  llvm::SmallVector<llvm::StringRef> option_strings;
+  llvm::SmallVector<llvm::StringRef> strings1;
+  llvm::SmallVector<llvm::StringRef> strings2;
+  llvm::SmallVector<llvm::StringRef> strings3;
+  auto parse = [&](llvm::ArrayRef<llvm::StringRef> args, llvm::raw_ostream& s) {
+    return Args::Parse(args, s, s, TestCommandInfo, [&](auto& b) {
+      b.AddStringOption({.name = "opt"},
+                        [&](auto& arg_b) { arg_b.Append(&option_strings); });
+      b.AddStringPositionalArg({.name = "s1"},
+                               [&](auto& arg_b) { arg_b.Append(&strings1); });
+      b.AddStringPositionalArg({.name = "s2"},
+                               [&](auto& arg_b) { arg_b.Append(&strings2); });
+      b.AddStringPositionalArg({.name = "s3"},
+                               [&](auto& arg_b) { arg_b.Append(&strings3); });
+      b.Do([] {});
+    });
+  };
+
+  EXPECT_THAT(parse({"a", "--opt=x", "b", "--opt=y", "--", "c", "--opt=z", "d",
+                     "--", "e", "f"},
+                    llvm::errs()),
+              Eq(Args::ParseResult::Success));
+  EXPECT_THAT(option_strings, ElementsAre(StrEq("x"), StrEq("y")));
+  EXPECT_THAT(strings1, ElementsAre(StrEq("a"), StrEq("b")));
+  EXPECT_THAT(strings2, ElementsAre(StrEq("c"), StrEq("--opt=z"), StrEq("d")));
+  EXPECT_THAT(strings3, ElementsAre(StrEq("e"), StrEq("f")));
+}
+
+TEST(ArgParserTest, OneOfOption) {
+  ssize_t value = 0;
+  auto parse = [&](llvm::ArrayRef<llvm::StringRef> args, llvm::raw_ostream& s) {
+    return Args::Parse(args, s, s, TestCommandInfo, [&](auto& b) {
+      b.AddOneOfOption({.name = "option"}, [&](auto& arg_b) {
+        arg_b.SetOneOf(
+            {
+                arg_b.OneOfValue("x", 1),
+                arg_b.OneOfValue("y", 2),
+                arg_b.OneOfValue("z", 3),
+            },
+            &value);
```

Comment:
I've been grappling with the OneOf template support. What would you think of an approach that only supports enums?

Reply:
Mainly I thought it might reduce the complexity. As noted elsewhere, I might want to iterate on the API so this seems fine for now.